In [4]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

# Importing Files

In [10]:
# setting the path
path = r'C:\Users\Bhati\OneDrive\Desktop\CF\Data Analytics Immersion\Achievement 4\11-Apr-2023 Instacart Basket Analysis'

In [11]:
# importing original data for product
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'),  index_col = False)

In [12]:
# importing prepared data for orders
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'),  index_col = False)

# Creating a dataframe with mix data types

In [13]:
# create a dataframe
df_test = pd.DataFrame()

In [80]:
# create a mixed type column
df_test['data'] = ['a','b',1,True] 

In [85]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [86]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [87]:
# Converting the data type to string
df_test['mix'] = df_test['mix'].astype('str')

In [88]:
# rechecking data type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col, ' mixed')
  else: print(col, ' consistent')

mix  consistent


# Checking for Missing values in Products

In [23]:
# to check for missing values in dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [28]:
# Subset of dataframe with Null values
df_Nan = df_prods[df_prods['product_name'].isnull() == True]

In [27]:
df_Nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [29]:
df_prods.shape

(49693, 5)

In [31]:
# creating new dataframe with no missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [33]:
df_prods_clean.shape

(49677, 5)

# Checking for duplicate values in Products

In [34]:
# Subset of dataframe with duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [35]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [36]:
df_prods_clean.shape

(49677, 5)

In [37]:
# Removing duplicate and storing in new dataframe
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [39]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting clean products file (after all data consistency checks)

In [47]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

Checking for anything that needs to be investigated further

In [58]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [59]:
df_prods_clean_no_dups[df_prods_clean_no_dups.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

Product_id 33664 with price of 99999 seems to be odd and could be a typo. Should be investigated further

# Consitency Checks for Orders Dataframe

In [60]:
# Checking for mixed data types in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
        print (col, ' mixed')
  else: print(col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
order_number  consistent
orders_day_of_week  consistent
order_hour_of_day  consistent
days_since_prior_order  consistent


In [67]:
# Checking for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [68]:
# create a dataframe for the missing values
df_ords_nan = df_ords [df_ords ['days_since_prior_order'].isnull () == True]

In [69]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

The reason for these missing values is that they might be first time customers, we can see order_number for these users is 1 and hence day_since_prior_order shows NaN

In [71]:
df_ords.shape

(3421083, 7)

6% of the total dataset is missing values. They are first time customers as we know that from order_number. Hence, at this point I will keep this dataset as the statistics could still be analysed.

In [72]:
# check for duplicate values in df_ords: 
df_ords_dups = df_ords[df_ords.duplicated()]

In [75]:
df_ords_dups.size

0

# Exporting clean order file (after all data consistency checks)

In [77]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))